packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -379,38 +393,87 @@ subtask(TASK_COMPILE_SOLIDITY_COMPILE_JOBS)
         return { artifactsEmittedPerJob: [] };
       }
 
-      // sort compilation jobs by compiler version
-      const sortedCompilationJobs = compilationJobs
-        .slice()
-        .sort((job1, job2) => {
-          return semver.compare(
-            job1.getSolcConfig().version,
-            job2.getSolcConfig().version
+      log(`Compiling ${compilationJobs.length} jobs`);
+
+      const versionList: string[] = [];
+      for (const job of compilationJobs) {
+        const solcVersion = job.getSolcConfig().version;
+
+        // versions older than 0.4.11 don't work with hardhat
+        // see issue https://github.com/nomiclabs/hardhat/issues/2004
+        if (semver.lt(solcVersion, COMPILE_TASK_FIRST_SOLC_VERSION_SUPPORTED)) {
+          throw new HardhatError(
+            ERRORS.BUILTIN_TASKS.COMPILE_TASK_UNSUPPORTED_SOLC_VERSION,
+            {
+              version: solcVersion,
+              firstSupportedVersion: COMPILE_TASK_FIRST_SOLC_VERSION_SUPPORTED,
+            }
           );
-        });
+        }
 
-      log(`Compiling ${sortedCompilationJobs.length} jobs`);
+        if (!versionList.includes(solcVersion)) {
+          versionList.push(solcVersion);
+        }
+      }
 
-      const artifactsEmittedPerJob: ArtifactsEmittedPerJob = [];
-      for (let i = 0; i < sortedCompilationJobs.length; i++) {
-        const compilationJob = sortedCompilationJobs[i];
+      /**
+       * I'm not super proud of this, but we have to download the list before the compilers,
+       * otherwise the next downloads fail with a race condition.
+       *
+       * :)
+       */
+      const compilersCache = await getCompilersDir();
+      const downloader = new CompilerDownloader(compilersCache);
+      await downloader.downloadCompilersList();
+
+      /**
+       * This is a pretty hacky solution to avoid a major refactor and/or breaking changes.
+       * The hack is the fact that we're just pre-downloading all of the necessary compilers
+       * before actually running the job, so all the later code assumes the compiler is already
+       * downloaded because if it wasn't then parallel compilation will break.
+       */
+      await chunkedPromiseAll(
```

Comment:
This should be done serially. Otherwise, there are some scenarios where the same version is downloaded at the same time and it results in an error.